App/HostedReactor.fs

```diff
@@ -0,0 +1,67 @@
+module HostedReactor
+
+open System
+open Microsoft.Extensions.Hosting
+open Propulsion.Internal
+open Propulsion.MessageDb
+
+type HostedReactor(connectionString, log, service: InvoiceNumberingReactor.Service, checkpointService: CheckpointStore.CheckpointService) =
+  inherit BackgroundService()
+
+  let stats =
+    { new Propulsion.Streams.Stats<_>(log, TimeSpan.FromMinutes 1, TimeSpan.FromMinutes 1) with
+        member _.HandleOk x = ()
+        member _.HandleExn(log, x) = () }
```

Comment:
why not  `log.Information(e, "Unhandled")` ?

Reply:
as noted in a different comment. Propulsion and its stats is an area I'm unfamiliar with. The only metrics I'm used to keeping are derived from spans and includes:

- Number of messages processed
- Cycle time
- Lead time
- Poll count (if this drops to 0 raise an alert)
- Lag from head of category
- Time lag from head of category

Reply:
In terms of generic stats:

A Streams-based Projector emits
- handler latency percentiles
- handler invocation counts
- event counts, unique streams counts
There are also state metrics such as longest duration a stream is having handler invocagtions without progress (if this exceeds SLA, raise alert)

Lag metrics are not a strong point
- Cosmos reader has some based on the CFP lag estimation
- Dynamo doesnt have anything much use
- Dynamo and others have a common feed reader base, which does emit metrics 

There are Serilog virtual sinks that parse the embedded metrics (which are embedded in the Serilog LogEvents), and emit those to the Prometheus sink)
In general any integration would curently be expected to work on that basis (observe the LogEvents, emit based on that)

(See the other comments for App level metrics, i.e recording/reporting outcomes from reactors in a way that Prometheus+Grafana can surface well is something that the dotnet-templates and the design of Propulsion in general tries to encourage)

The lowest common denominator is also a key part though - a Reactor or Projector should, without requiring a suite of Dockerized stuff to be configured or learned, give plenty clues to what's going on in good and bad times - error rates, throughput and latency.
This even extends to test scenarios - a test run should dump stats as a Reactor invocation on the console would. (There is wiring for this in https://github.com/jet/dotnet-templates/pull/127 and equinox-shipping and more; I've yet to to the wiring for Propulsion.MessageDb to be used as a source in that sample, but the Stats are managed in the same way regardless of the source)